### PR TITLE
Add a helper function to compare objects in various ways: shallow, deep, partial

### PR DESCRIPTION
**Pull request by jsnajdr**
Module that exports a `compareProps( options )` function that returns an object comparator with the desired properties.

The default is to compare everything shallowly and there are options to compare selected properties deeply or ignore them completely.

Very useful when optimizing React components and Redux `connect` calls by adding `shouldComponentUpdate` or `areStatePropsEqual`. I extracted this module from my WIP work that optimizes 10+ components that benefit from it.

See the README.md file supplied for documentation and the tests for examples.

The implementation is strongly inspired by the [is-equal-shallow](https://github.com/jonschlinkert/is-equal-shallow) package used internally by React to implement `PureComponent`. Very simple and
fast code that doesn't do any array allocations (no `Object.keys`, `map` or `filter`) or other fancy stuff.

### CODE REVIEW

File: client/lib/compare-props/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * External dependencies
+ */
+
+import { includes, isEqual } from 'lodash';
+
+export default ( options = {} ) => {
+	const { ignore, deep, shallow } = options;
+	return ( prevProps, nextProps ) => {
```
Comment (lamosty) [inline client/lib/compare-props/index.js]: This sounds React-ish. Maybe name it `value`, `other` akin to Lodash?

File: client/lib/compare-props/index.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			// Compare all other props (or a selected subset) shallowly
+			if ( ! shallow || includes( shallow, propName ) ) {
+				if ( prevProps[ propName ] !== nextProps[ propName ] ) {
+					return false;
+				}
+			}
+		}
+
+		// Find properties that are only in `nextProps` and are not ignored.
+		// Presence of such properties means that the object are not equal.
```
Comment (lamosty) [inline client/lib/compare-props/index.js]: "  that the objectS are not equal." typo.

**Comment by samouri** [comment]
This seems pretty useful. Can you point to the WIP PR so I can see how it would be used?

File: client/lib/compare-props/test/index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+					page: 2,
+				},
+				{
+					query: { text: 'plugin' },
+					page: 2,
+				}
+			)
+		).to.be.true;
+
+		expect(
```
Comment (samouri) [inline client/lib/compare-props/test/index.js]: This tests that everything not specified as deep stays as a shallow compare right?
Whats your take on cases like this where it feels like the unit test might be exceeding the bounds of the description?

The real description should be: 'should compare selected properties deeply while remaining properties are compared shallowly'

File: client/lib/compare-props/test/index.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		const comparator = compareProps( { shallow: [ 'id' ] } );
+
+		expect( comparator( { id: 1, is_jetpack: true }, { id: 1, is_domain_only: true } ) ).to.be.true;
+
+		expect( comparator( { is_jetpack: true }, { id: 1, is_domain_only: true } ) ).to.be.false;
+
+		expect( comparator( { id: 1, is_jetpack: true }, { is_domain_only: true } ) ).to.be.false;
+	} );
+
+	test( 'should prefer `ignore` list over `deep` and both over `shallow`', () => {
```
Comment (samouri) [inline client/lib/compare-props/test/index.js]: this one is cool

**Comment by jsnajdr** [comment]
> This seems pretty useful. Can you point to the WIP PR so I can see how it would be used?

I think you independently discovered one possible usage of this function today, namely the `StatsModuleChartTabs` component.

There the shallow comparison always fails on the `quickQuery` and `fullQuery` props, as they are new objects each time. And it's hard to make them identical and shared across calls.

I'd fix that by adding a `areStatePropsEqual` option to its `connect`:
```js
connect(
  mapStateToProps,
  mapDispatchToProps,
  null,
  {
    areStatePropsEqual: compareProps( { deep: [ 'quickQuery', 'fullQuery' ] } )
  }
)( StatsModuleChartTab );
```

File: client/lib/compare-props/index.js (hunk 1 above)
Comment (jsnajdr) [reply client/lib/compare-props/index.js]: The purpose if this function is to be used with React props. There are many ways how to compare Javascript objects and the ambition here is not to implement some generic object comparison. For example, comparing objects with unusual prototype chains would probably not always work as expected. But for React props, it doesn't matter.

File: client/lib/compare-props/index.js (hunk 2 above)
Comment (jsnajdr) [reply client/lib/compare-props/index.js]: You read everything! Fixed :)

File: client/lib/compare-props/test/index.js (hunk 3 above)
Comment (jsnajdr) [reply client/lib/compare-props/test/index.js]: I updated the test description.

**Comment by dmsnell** [comment]
Awesome! I too know some components where this would be useful, especially in places where our use of `moment` is screwing things up (a timestamp is converted to a _new_ moment object on every `connect()` call). Actually I think there are often other fixes that are better and don't require this, but those fixes aren't always easy or accessible.

**Pull request merged**